src/txrequest.h

```diff
@@ -0,0 +1,341 @@
+        //! The CANDIDATE_BEST is the lowest-priority entry among all CANDIDATE_OTHER (and _BEST) ones for that txid.
+        CANDIDATE_BEST,
+        //! A REQUESTED entry.
+        REQUESTED,
+        //! A CANDIDATE entry that's not CANDIDATE_NEW or CANDIDATE_BEST.
+        CANDIDATE_OTHER,
+        //! A COMPLETED entry.
+        COMPLETED,
+    };
+
+    //! Tag for the EntryPeer-based index.
+    struct ByPeer {};
+    //! Tag for the EntryTxid-based index.
+    struct ByTxid {};
+    //! Tag for the EntryTime-based index.
+    struct ByTime {};
+
+    //! The ByPeer index is sorted by (peer, state == CANDIDATE_BEST, txid)
+    using EntryPeer = std::tuple<uint64_t, bool, const uint256&>;
+
+    //! The ByTxid index is sorted by (txid, state, priority [CANDIDATE_OTHER]; 0 [otherwise])
+    using EntryTxid = std::tuple<const uint256&, State, uint64_t>;
+
+    //! The ByTime index is sorted by (0 [CANDIDATE_NEW,REQUESTED]; 1 [COMPLETED];
+    // 2 [CANDIDATE_OTHER,CANDIDATE_BEST], time)
+    using EntryTime = std::pair<int, std::chrono::microseconds>;
+
+    //! The current sequence number. Increases for every announcement. This is used to sort txids returned by
+    //! GetRequestable in announcement order.
+    uint64_t m_sequence{0};
+
+    //! An announcement entry.
+    struct Entry {
+        //! Txid that was announced.
+        const uint256 m_txid;
+        //! For CANDIDATE{OTHER,_DELAYED,_BEST} the reqtime; for REQUESTED the exptime
+        std::chrono::microseconds m_time;
+        //! What peer the request was from.
+        const uint64_t m_peer;
+        //! What sequence number this announcement has.
+        const uint64_t m_sequence : 59;
+        //! Whether it's from an outbound peer or not.
+        const bool m_inbound : 1;
+        //! What state this announcement is in.
+        State m_state : 3;
+        //! Whether this was the very first announcement for this txid.
+        //! It's only modified right after insertion, at which point it doesn't affect position
+        //! in any of the indexes. Make it mutable to enable changing without triggering reindex.
+        mutable bool m_first : 1;
+
+        //! Whether this entry is selected. There can be at most 1 selected peer per txid.
+        bool IsSelected() const { return m_state == State::CANDIDATE_BEST || m_state == State::REQUESTED; }
+
+        //! Whether this entry is waiting for a certain time to pass.
+        bool IsWaiting() const { return m_state == State::REQUESTED || m_state == State::CANDIDATE_NEW; }
+
+        //! Whether this entry can feasibly be selected if the current IsSelected() one disappears.
+        bool IsSelectable() const { return m_state == State::CANDIDATE_OTHER || m_state == State::CANDIDATE_BEST; }
+
+        //! Construct a new entry from scratch
+        Entry(const uint256& txid, uint64_t peer, bool inbound, State state, std::chrono::microseconds time, uint64_t
```

Comment:
I'm trying not to leave too many nits at this stage, but what kind of maniac breaks lines between the type and parameter name?!

Reply:
🤪

Reply:
Fixed.